Prefer videoBinned camera presets in case resizing is needed

youennf opened the pull request:
#### 3664d7d244e0924b68bd6fd9b24f0f6aa3a7821a
<pre>
Prefer videoBinned camera presets in case resizing is needed
<a href="https://bugs.webkit.org/show_bug.cgi?id=275317">https://bugs.webkit.org/show_bug.cgi?id=275317</a>
<a href="https://rdar.apple.com/129506159">rdar://129506159</a>

Reviewed by Eric Carlson.

In case we select a preset that might require downsampling, it is good to favor binned presets as this is better for power efficiency.
Previously, we were favoring videoBinned presets if they were closer to the final size.

We can be more aggressive than that and favor videoBinned presets even if they are further from the final size.
We still try to select the smallest possible videoBinned preset.
This is implemented in RealtimeVideoCaptureSource.

As video binned format may not always be the best preset, we use the powerEfficientPixelFormat as a constraint that can be given to getUserMedia/applyConstraints
to aggressively select binned presets.
We do not expose settings and capabilities for this constraint since we only use it as a hint that the web application is wanting power efficiency.

We update some code to allow testing this feature.
We add a isEfficient field to VideoPreset so that we can mock efficient presets.
We have one such mock preset (environment camera, 1920 x 1080) that we use for testing.

We add the infrastructure to check whether a given capture track is power efficient or not.

We have to update LayoutTests/fast/mediastream/getUserMedia-video-rescaling.html given it will now favor 1920x1080 preset, which is 16/9.

* LayoutTests/fast/mediastream/MediaDevices-getSupportedConstraints-expected.txt:
* LayoutTests/fast/mediastream/camera-powerEfficient-track-expected.txt: Added.
* LayoutTests/fast/mediastream/camera-powerEfficient-track.html: Added.
* LayoutTests/fast/mediastream/getUserMedia-video-rescaling.html:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::hasInvalidGetDisplayMediaConstraint):
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.cpp:
(WebCore::convertToInternalForm):
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.idl:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.idl:
* Source/WebCore/platform/mediastream/MediaConstraintType.cpp:
(WebCore::convertToString):
* Source/WebCore/platform/mediastream/MediaConstraintType.h:
* Source/WebCore/platform/mediastream/MediaConstraints.cpp:
(WebCore::MediaTrackConstraintSetMap::filter const):
(WebCore::MediaTrackConstraintSetMap::set):
(WebCore::MediaTrackConstraintSetMap::merge):
(WebCore::MediaTrackConstraintSetMap::isValid const):
(WebCore::MediaConstraints::setDefaultVideoConstraints):
(WebCore::MediaTrackConstraintSetMap::isolatedCopy const):
* Source/WebCore/platform/mediastream/MediaConstraints.h:
(WebCore::MediaTrackConstraintSetMap::MediaTrackConstraintSetMap):
(WebCore::MediaTrackConstraintSetMap::powerEfficientPixelFormat const):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::fitnessDistance):
(WebCore::RealtimeMediaSource::applyConstraint):
(WebCore::RealtimeMediaSource::selectSettings):
(WebCore::RealtimeMediaSource::supportsConstraint):
(WebCore::RealtimeMediaSource::hasAnyInvalidConstraint):
(WebCore::RealtimeMediaSource::extractVideoPresetConstraints):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.cpp:
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsConstraint const):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::shouldUsePreset):
(WebCore::computePresetToUse):
(WebCore::RealtimeVideoCaptureSource::bestSupportedSizeFrameRateAndZoom):
(WebCore::RealtimeVideoCaptureSource::isPowerEfficient const):
(WebCore::RealtimeVideoCaptureSource::shouldUsePreset): Deleted.
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/VideoPreset.h:
(WebCore::VideoPreset::VideoPreset):
(WebCore::VideoPreset::isEfficient const):
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::isFormatPowerEfficient):
(WebCore::AVVideoCaptureSource::generatePresets):
(WebCore::AVVideoCaptureSource::prefersPreset): Deleted.
* Source/WebCore/platform/mock/MockMediaDevice.h:
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::defaultDevices):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isMediaStreamTrackPowerEfficient const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::dispatchSyncMessage):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::isPowerEfficient):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::isPowerEfficient const):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:

Canonical link: <a href="https://commits.webkit.org/280024@main">https://commits.webkit.org/280024@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/30527e3cfe94a355533f489b4e6a5f99fa71042d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55536 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34859 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8003 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58520 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5966 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57662 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42481 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6165 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44721 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4096 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57565 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/32765 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47864 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25849 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29549 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5192 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4110 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5460 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60111 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30690 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5583 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/52155 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31775 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47934 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/51638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12305 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32856 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31522 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
```diff
@@ -225,9 +222,21 @@ bool RealtimeVideoCaptureSource::supportsCaptureSize(std::optional<int> width, s
+static bool shouldUsePreset(const VideoPreset& current, const VideoPreset& candidate, bool shouldPreferPowerEfficiency)
 {
-    return candidate.size().width() <= current.size().width() && candidate.size().height() <= current.size().height() && prefersPreset(candidate);
+    if (shouldPreferPowerEfficiency && candidate.isEfficient() && !current.isEfficient())
+        return true;
```
If more than one preset is power efficient, should we prefer the one that is closest to the desired size?

youennf replied:
I think this is what next line will do

eric-carlson commented on Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
```diff
@@ -934,6 +938,11 @@ void RealtimeMediaSource::applyConstraint(MediaConstraintType constraintType, co
     }
+    case MediaConstraintType::PowerEfficientPixelFormat: {
+        ASSERT(constraint.isBoolean());
+        // FIXME: Add support
```
Nit: it would be good to have a bug filed for this and `BackgroundBlur`

youennf committed:
Prefer videoBinned camera presets in case resizing is needed
https://bugs.webkit.org/show_bug.cgi?id=275317
rdar://129506159

Reviewed by Eric Carlson.

In case we select a preset that might require downsampling, it is good to favor binned presets as this is better for power efficiency.
Previously, we were favoring videoBinned presets if they were closer to the final size.

We can be more aggressive than that and favor videoBinned presets even if they are further from the final size.
We still try to select the smallest possible videoBinned preset.
This is implemented in RealtimeVideoCaptureSource.

As video binned format may not always be the best preset, we use the powerEfficientPixelFormat as a constraint that can be given to getUserMedia/applyConstraints
to aggressively select binned presets.
We do not expose settings and capabilities for this constraint since we only use it as a hint that the web application is wanting power efficiency.

We update some code to allow testing this feature.
We add a isEfficient field to VideoPreset so that we can mock efficient presets.
We have one such mock preset (environment camera, 1920 x 1080) that we use for testing.

We add the infrastructure to check whether a given capture track is power efficient or not.

We have to update LayoutTests/fast/mediastream/getUserMedia-video-rescaling.html given it will now favor 1920x1080 preset, which is 16/9.

* LayoutTests/fast/mediastream/MediaDevices-getSupportedConstraints-expected.txt:
* LayoutTests/fast/mediastream/camera-powerEfficient-track-expected.txt: Added.
* LayoutTests/fast/mediastream/camera-powerEfficient-track.html: Added.
* LayoutTests/fast/mediastream/getUserMedia-video-rescaling.html:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::hasInvalidGetDisplayMediaConstraint):
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.cpp:
(WebCore::convertToInternalForm):
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.idl:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.idl:
* Source/WebCore/platform/mediastream/MediaConstraintType.cpp:
(WebCore::convertToString):
* Source/WebCore/platform/mediastream/MediaConstraintType.h:
* Source/WebCore/platform/mediastream/MediaConstraints.cpp:
(WebCore::MediaTrackConstraintSetMap::filter const):
(WebCore::MediaTrackConstraintSetMap::set):
(WebCore::MediaTrackConstraintSetMap::merge):
(WebCore::MediaTrackConstraintSetMap::isValid const):
(WebCore::MediaConstraints::setDefaultVideoConstraints):
(WebCore::MediaTrackConstraintSetMap::isolatedCopy const):
* Source/WebCore/platform/mediastream/MediaConstraints.h:
(WebCore::MediaTrackConstraintSetMap::MediaTrackConstraintSetMap):
(WebCore::MediaTrackConstraintSetMap::powerEfficientPixelFormat const):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::fitnessDistance):
(WebCore::RealtimeMediaSource::applyConstraint):
(WebCore::RealtimeMediaSource::selectSettings):
(WebCore::RealtimeMediaSource::supportsConstraint):
(WebCore::RealtimeMediaSource::hasAnyInvalidConstraint):
(WebCore::RealtimeMediaSource::extractVideoPresetConstraints):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.cpp:
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsConstraint const):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::shouldUsePreset):
(WebCore::computePresetToUse):
(WebCore::RealtimeVideoCaptureSource::bestSupportedSizeFrameRateAndZoom):
(WebCore::RealtimeVideoCaptureSource::isPowerEfficient const):
(WebCore::RealtimeVideoCaptureSource::shouldUsePreset): Deleted.
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/VideoPreset.h:
(WebCore::VideoPreset::VideoPreset):
(WebCore::VideoPreset::isEfficient const):
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::isFormatPowerEfficient):
(WebCore::AVVideoCaptureSource::generatePresets):
(WebCore::AVVideoCaptureSource::prefersPreset): Deleted.
* Source/WebCore/platform/mock/MockMediaDevice.h:
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::defaultDevices):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isMediaStreamTrackPowerEfficient const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::dispatchSyncMessage):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::isPowerEfficient):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::isPowerEfficient const):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:

Canonical link: https://commits.webkit.org/280024@main
